awx/main/scheduler/task_manager.py

```diff
@@ -122,19 +122,51 @@ def job_blocked_by(self, task):
+        tasks = []
+        now = tz_now()
+
+        # Get query sets for all jobs. In the future we'll just query UnifiedJob
+        jobs_qs = Job.objects.filter(status__in=status_list).prefetch_related('instance_group').order_by('modified')[:limit]
         inventory_updates_qs = (
-            InventoryUpdate.objects.filter(status__in=status_list).exclude(source='file').prefetch_related('inventory_source', 'instance_group')
+            InventoryUpdate.objects.filter(status__in=status_list)
+            .exclude(source='file')
+            .prefetch_related('inventory_source', 'instance_group')
+            .order_by('modified')[:limit]
         )
-        inventory_updates = [i for i in inventory_updates_qs]
         # Notice the job_type='check': we want to prevent implicit project updates from blocking our jobs.
-        project_updates = [p for p in ProjectUpdate.objects.filter(status__in=status_list, job_type='check').prefetch_related('instance_group')]
-        system_jobs = [s for s in SystemJob.objects.filter(status__in=status_list).prefetch_related('instance_group')]
-        ad_hoc_commands = [a for a in AdHocCommand.objects.filter(status__in=status_list).prefetch_related('instance_group')]
-        workflow_jobs = [w for w in WorkflowJob.objects.filter(status__in=status_list)]
-        all_tasks = sorted(jobs + project_updates + inventory_updates + system_jobs + ad_hoc_commands + workflow_jobs, key=lambda task: task.created)
+        project_updates_qs = (
+            ProjectUpdate.objects.filter(status__in=status_list, job_type='check').prefetch_related('instance_group').order_by('modified')[:limit]
+        )
+        system_jobs_qs = SystemJob.objects.filter(status__in=status_list).prefetch_related('instance_group').order_by('modified')[:limit]
+        ad_hoc_commands_qs = AdHocCommand.objects.filter(status__in=status_list).prefetch_related('instance_group').order_by('modified')[:limit]
+        workflow_jobs_qs = WorkflowJob.objects.filter(status__in=status_list).order_by('modified')[:limit]
+
+        # unpack query sets to a list because that is the return type expected
+        for task_qs in [jobs_qs, inventory_updates_qs, project_updates_qs, system_jobs_qs, ad_hoc_commands_qs, workflow_jobs_qs]:
+            tasks.extend([t for t in task_qs])
+
+        tasks = sorted(tasks, key=lambda task: task.created)
+
+        # update modified on pending tasks so we loop over all pending tasks in a window "limit" size, making sure we don't get stuck re-processing
+        # possibly unstartable pending tasks. E.g. we get the least recently modified each time, but each time we update modified so the next
+        # round of the task manager will pick up a new set of pending jobs (in the case that there are jobs stuck in pending due to capacity or other issues)
+        if status_list == ['pending'] and tasks:
+            UnifiedJob.objects.non_polymorphic().filter(pk__in=[t.pk for t in tasks]).update(modified=now)
```

Comment:
This is a pretty heavy update here. In severe situations, we're talking about 100 updates due to the task start limit. This is committed as a part of the overall task manager transaction, which can take up to 5 minutes, but commonly in the 10s of seconds. I'm less concerned about the number of 100, as opposed to the fact that it will happen over and over again to the same jobs. The task manager will run way more than once per 20 seconds in heavy load, and this will result in a gob-smacking amount of activity.

It also seems at least half-way unnecessary. Let's say we go into this with 30 pending jobs that are blocked on a resource or capacity, and 30 pending jobs that will be transitioned here. This will update the `modified` of all 60 jobs, but only 30 of them were necessary to update for your goal. It would be possible to, instead, track the blocked / un-started pending jobs and make the modified update at the end, meaning only 30 rows would need to be updated.

This also kind of violates our ideas of processing order. We sort tasks in order of `created`, but if (number of pending jobs) > (task start limit) it won't really be functioning that way. Jobs started later will jump the queue in front of other jobs started earlier.

I dislike the direction this is going as well. I want us to be doing fewer things in the task manager, this adds more, and it will be hard to get rid of.

Reply:
dropped the update